### PR TITLE
Initial Release

### PR DESCRIPTION
**Pull request by shiva-cirus**
Initial Release of Amazon SQS Streaming source and Batch sink plugins.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+      return;
+    }
+
+    if (Utility.isNullOrEmpty(schema)) {
+      collector.addFailure("Schema must be specified.", null)
+        .withConfigProperty(PROPERTY_SCHEMA);
+    }
+  }
+
+  private void validateFormat(FailureCollector collector) {
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:382]: Seems like `validateSchema()` can be combined with this function, they seem tightly coupled. Is there a use case for having them separate?

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    }
+
+    // if format is empty, there must be just a single message field of type bytes or nullable
+    // types.
+    if (Utility.isNullOrEmpty(format)) {
+      List<Schema.Field> messageFields = messageSchema.getFields();
+      if (messageFields.size() > 1) {
+        List<String> fieldNames = messageFields.stream().map(o -> o.getName()).collect(Collectors.toList());
+        collector.addFailure(String.format("Without a format, the schema must contain just a single message field " +
+            "of type bytes or nullable bytes. Found %s message fields (%s).", messageFields.size(),
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:407]: Should be `%d` for `messageFields.size()`

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+          Joiner.on(',').join(fieldNames)), null)
+          .withConfigProperty(PROPERTY_SCHEMA);
+        return;
+      }
+      Schema.Field messageField = messageFields.get(0);
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:412]: What if the schema has 0 fields? We should check that before getting the first element

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+      }
+    } else {
+      // otherwise, if there is a format, make sure we can instantiate it.
+      FormatSpecification formatSpec =
+        new FormatSpecification(format, messageSchema, new HashMap<String, String>());
+
+      try {
+        RecordFormats.createInitializedFormat(formatSpec);
+      } catch (Exception e) {
+        collector.addFailure(String.format("Unable to instantiate a message parser from format '%s' and message " +
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:432]: Can we make this error more user-friendly? This doesnt given the user clear next steps or even a clear idea what the issue is. Maybe something like :

> Current schema ([SCHEMA]) does not match the format expected for [FORMAT]: [ERROR]

Are here any common errors or corrective steps we can recommend?

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+      }
+
+      @Override
+      public void onStart() {
+        new Thread() {
+          @Override
+          public void run() {
+            AmazonSQS sqs = null;
+            try {
+              sqs = SqsUtil.getSqsClient(conf.getAuthenticationMethod(), conf.getAccessId(),
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java:80]: Do we have to shutdown this client on stop?

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+          return Collections.emptyList();
+        }
+
+        String queueName = config.getQueueName();
+        String queueUrl = sqs.getQueueUrl(queueName).getQueueUrl();
+        // Enable long polling on an existing queue
+        SetQueueAttributesRequest setQueueAttributesRequest = new SetQueueAttributesRequest()
+          .withQueueUrl(queueUrl)
+          .addAttributesEntry("ReceiveMessageWaitTimeSeconds", String.valueOf(config.getWaitTime()));
+        sqs.setQueueAttributes(setQueueAttributesRequest);
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java:133]: Does this have to be set every time we fetch messages or can it just be done when we create the client?

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            } catch (SqsInitializationException e) {
+              throw new RuntimeException(e);
+            }
+
+            while (!isStopped()) {
+              try {
+                List<Message> messages = fetchMessages(sqs, conf);
+                LOG.debug("messages.size() = {}", messages.size());
+                for (Message message : messages) {
+                  store(toRecord(message, conf));
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java:91]: How does `store()` handle null values? This logic implies it just ignores them, is that correct?

File: src/main/java/io/cdap/plugin/aws/sqs/source/util/SqsSourceConstants.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  String PROPERTY_FORMAT = "format";
+
+  /**
+   * Configuration property name used to specify the schema.
+   */
+  String PROPERTY_SCHEMA = "schema";
+
+  int MAX_WAIT_TIME = 20;
+  int MAX_MESSAGES_LIMIT = 10;
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/source/util/SqsSourceConstants.java:55]: Comments for these constants?

File: src/test/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfigTest.java [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+  @Test
+  public void testAuthMethodIAM() {
+    SqsSinkConfig config = SqsSinkConfigHelper.newConfigBuilder()
+      .setAuthenticationMethod("IAM")
+      .build();
+
+    try {
+      MockFailureCollector collector = new MockFailureCollector();
+      config.getAuthenticationMethod(collector);
```
Comment (MEseifan) [inline src/test/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfigTest.java:73]: Instead of throwing the error in `getAuthenticationMethod` and catching it, you can check the contents of `MockFailureCollector` to see if the error was generated properly. This comment applies to all of the validate methods

File: src/test/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfigTest.java [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+      collector.getOrThrowException();
+    } catch (ValidationException e) {
+      Assert.assertEquals(1, e.getFailures().size());
+      Assert.assertEquals(PROPERTY_SCHEMA, e.getFailures().get(0).getCauses().get(0)
+        .getAttribute(CauseAttributes.STAGE_CONFIG));
+    }
+  }
+
+  @Test
+  public void testValidateFormatInvalidSchema() {
```
Comment (MEseifan) [inline src/test/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfigTest.java:370]: Maybe add a test for a valid but empty schema?

File: widgets/sqs-streamingsource.json [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+          }
+        }
+      ]
+    },
+    {
+      "label": "Advanced",
+      "properties": [
+        {
+          "widget-type": "select",
+          "label": "Delete messages after read?",
```
Comment (MEseifan) [inline widgets/sqs-streamingsource.json:107]: Consider using a toggle for true/false values:

```
{
    "widget-type": "toggle",
    "name": "delete",
    "label": "Delete messages after read?",
    "widget-attributes": {
        "default": "true",
        "on": {
            "value": "true",
            "label": "Yes"
        },
        "off": {
            "value": "false",
            "label": "No"
        }
    }
},
```

File: widgets/sqs-streamingsource.json [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+            "default": "true"
+          }
+        },
+        {
+          "widget-type": "select",
+          "label": "Format",
+          "name": "format",
+          "widget-attributes": {
+            "values": [
+              "",
```
Comment (MEseifan) [inline widgets/sqs-streamingsource.json:123]: Why not add `bytes` as an option? That seems to be what `""` implicitly means in the backend, might avoid some confusion for users

File: checkstyle.xml [hunk 13]
```diff
@@ -0,0 +1,405 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright © 2014-2015 Cask Data, Inc.
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply checkstyle.xml]: Done

File: docs/sqs-streamingsource.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
+# Amazon SQS Source
+
+Description
+-----------
+
+The source reads messages from SQS Queue using the long polling option of the API. SQS currently does provide ability 
+for call backs or streaming. The plugin will continuously issues the receive message call of the api to read messages
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply docs/sqs-streamingsource.md]: Done

File: docs/sqs-streamingsource.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+**Delete messages after read?** Permanently delete messages from the queue after successfully reading. If set to 
+`false`, the messages will stay on queue and will be received again leading to duplicates.
+
+**Interval**: The amount of time to wait between each poll in seconds. The plugin will wait for the duration specified 
+before issuing an receive message call.
+
+**Wait Time**: The duration (in seconds) for which the call waits for a message to arrive in the queue before 
+returning. This will be set as part of the ReceiveMessage API call action. The maximum wait time currently allowed by 
+SQS api 20 seconds.
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply docs/sqs-streamingsource.md]: Done

File: pom.xml [hunk 16]
```diff
@@ -0,0 +1,442 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright © 2016-2019 Cask Data, Inc.
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply pom.xml]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/exception/SqsInitializationException.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/exception/SqsInitializationException.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSink.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    this.sqsSinkConfig = sqsSinkConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    StageConfigurer configurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector collector = configurer.getFailureCollector();
+    sqsSinkConfig.validate(collector);
+    collector.getOrThrowException();
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSink.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSink.java [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+        .map(Schema.Field::getName)
+        .collect(Collectors.toList()));
+
+  }
+
+  private List<Object> getExtractedValues(StructuredRecord input, List<Schema.Field> fields) {
+    // Extract all values from the structured record
+    List<Object> objs = Lists.newArrayList();
+    for (Schema.Field field : fields) {
+      objs.add(input.get(field.getName()).toString());
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSink.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSink.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+    String body;
+    if (sqsSinkConfig.getMessageFormat() == SinkMessageFormat.JSON) {
+      body = StructuredRecordStringConverter.toJsonString(input);
+    } else {
+      List<Object> objs = getExtractedValues(input, fields);
+      body = StringUtils.join(objs, ",");
+    }
+
+    emitter.emit(new KeyValue<>((Text) null, new Text(body)));
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSink.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+  }
+
+  public SqsAuthMethod getAuthenticationMethod(FailureCollector collector) {
+    SqsAuthMethod authMethod = getAuthenticationMethod();
+    if (authMethod == SqsAuthMethod.ACCESS_CREDENTIALS) {
+      return authMethod;
+    }
+
+    if (authMethod == SqsAuthMethod.IAM) {
+      collector.addFailure(String.format("{} auth method is not implemented", authMethod), null)
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    String supportedRegions = Arrays.stream(Regions.values()).map(Regions::getName)
+      .collect(Collectors.joining(", "));
+
+    collector.addFailure("Unsupported region value: " + region,
+      String.format("Supported regions are: %s", supportedRegions))
+      .withConfigProperty(PROPERTY_REGION);
+
+    throw collector.getOrThrowException();
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+    if (authMethod == SqsAuthMethod.IAM) {
+      collector.addFailure(String.format("{} auth method is not implemented", authMethod), null)
+        .withConfigProperty(PROPERTY_AUTH_METHOD);
+    } else {
+      collector.addFailure("Unsupported auth method value: " + authenticationMethod,
+        String.format("Supported auth methods are: %s", SqsAuthMethod.getSupportedAuthMethods()))
+        .withConfigProperty(PROPERTY_AUTH_METHOD);
+    }
+
+    throw collector.getOrThrowException();
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+  public SinkMessageFormat getMessageFormat(FailureCollector collector) {
+    SinkMessageFormat format = getMessageFormat();
+    if (format != null) {
+      return format;
+    }
+
+    collector.addFailure("Unsupported format value: " + messageFormat,
+      String.format("Supported formats are: %s", SinkMessageFormat.getSupportedFormats()))
+      .withConfigProperty(PROPERTY_MESSAGE_FORMAT);
+    throw collector.getOrThrowException();
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+          .withConfigProperty(PROPERTY_ACCESS_ID);
+      }
+
+      if (Utility.isNullOrEmpty(accessKey)) {
+        collector.addFailure("Access Key must be specified.", null)
+          .withConfigProperty(PROPERTY_ACCESS_KEY);
+      }
+    }
+
+    validateSqsConnection(collector);
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: accessId and accessKey will be empty in case of AuthMode = IAM

File: src/main/java/io/cdap/plugin/aws/sqs/sink/util/SqsSinkConstants.java [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+   * Configuration property name used to specify delay in seconds
+   */
+  String PROPERTY_DELAY_SECONDS = "delaySeconds";
+
+  /**
+   * Configuration property name used to specify message format
+   */
+  String PROPERTY_MESSAGE_FORMAT = "messageFormat";
+
+  int MAX_DELAY = 900;
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/util/SqsSinkConstants.java:37]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSource.java [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    StageConfigurer configurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector collector = configurer.getFailureCollector();
+    config.validate(collector);
+    collector.getOrThrowException();
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSource.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 28]
```diff
@@ -0,0 +1,440 @@
...
+  }
+
+  public SqsAuthMethod getAuthenticationMethod(FailureCollector collector) {
+    SqsAuthMethod authMethod = getAuthenticationMethod();
+    if (authMethod == SqsAuthMethod.ACCESS_CREDENTIALS) {
+      return authMethod;
+    }
+
+    if (authMethod == SqsAuthMethod.IAM) {
+      collector.addFailure(String.format("{} auth method is not implemented", authMethod), null)
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 29]
```diff
@@ -0,0 +1,440 @@
...
+import static io.cdap.plugin.aws.sqs.util.SqsConstants.PROPERTY_ACCESS_KEY;
+import static io.cdap.plugin.aws.sqs.util.SqsConstants.PROPERTY_AUTH_METHOD;
+import static io.cdap.plugin.aws.sqs.util.SqsConstants.PROPERTY_QUEUE_NAME;
+import static io.cdap.plugin.aws.sqs.util.SqsConstants.PROPERTY_REGION;
+import static io.cdap.plugin.aws.sqs.util.SqsConstants.PROPERTY_SQS_ENDPOINT;
+
+/**
+ * Configuration for the {@link SqsSource}.
+ */
+public class SqsSourceConfig extends ReferencePluginConfig implements Serializable {
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 30]
```diff
@@ -0,0 +1,440 @@
...
+          .withConfigProperty(PROPERTY_ACCESS_ID);
+      }
+
+      if (Utility.isNullOrEmpty(accessKey)) {
+        collector.addFailure("Access Key must be specified.", null)
+          .withConfigProperty(PROPERTY_ACCESS_KEY);
+      }
+    }
+
+    validateSqsConnection(collector);
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java [hunk 31]
```diff
@@ -0,0 +1,294 @@
...
+  }
+
+  @VisibleForTesting
+  void validateSqsConnection(FailureCollector collector) {
+    try {
+      AmazonSQS sqs = SqsUtil.getSqsClient(getAuthenticationMethod(), accessId, accessKey, sqsEndpoint, region,
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: We did not notice slowness so we will keep it in the validation path

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java [hunk 32]
```diff
@@ -0,0 +1,440 @@
...
+  }
+
+  private void validateMaxNumberOfMessages(FailureCollector collector) {
+    if (containsMacro(PROPERTY_NUMBER_OF_MESSAGES)) {
+      return;
+    }
+
+    if (numberOfMessages < 1 || numberOfMessages > MAX_MESSAGES_LIMIT) {
+      collector.addFailure(String.format("Invalid number of messages '%d'.", numberOfMessages),
+        String.format("Ensure the number of messages is 1 or at most '%d'", MAX_MESSAGES_LIMIT))
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java (hunk 1 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java (hunk 2 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java (hunk 3 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:247]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java (hunk 4 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java (hunk 5 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java:80]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java (hunk 6 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java (hunk 7 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java:100]: As you pointed out it is handling null values by ignoring and writing an error log with details. Can you please how do we handle this case more effectively?

File: src/main/java/io/cdap/plugin/aws/sqs/source/util/SqsSourceConstants.java (hunk 8 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/aws/sqs/source/util/SqsSourceConstants.java:62]: Done

File: src/test/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfigTest.java (hunk 9 above)
Comment (amolkokane76) [reply src/test/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfigTest.java:62]: Done

File: src/test/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfigTest.java (hunk 10 above)
Comment (amolkokane76) [reply src/test/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfigTest.java:356]: Done

File: widgets/sqs-streamingsource.json (hunk 11 above)
Comment (amolkokane76) [reply widgets/sqs-streamingsource.json]: Done

File: widgets/sqs-streamingsource.json (hunk 12 above)
Comment (amolkokane76) [reply widgets/sqs-streamingsource.json:127]: Done

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (MEseifan) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: Got it, makes sense then

File: src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (MEseifan) [reply src/main/java/io/cdap/plugin/aws/sqs/sink/SqsSinkConfig.java]: noted

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java (hunk 7 above)
Comment (MEseifan) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceUtil.java:100]: If it ignores the null values then that should be fine. I just wanted to make sure those werent being stored.

File: widgets/sqs-streamingsource.json (hunk 12 above)
Comment (MEseifan) [reply widgets/sqs-streamingsource.json:127]: Do we still need "" as an option? Can we set the default value to "bytes"?

File: src/main/java/io/cdap/plugin/aws/sqs/common/SqsPluginConfig.java [hunk 33]
```diff
@@ -0,0 +1,263 @@
...
+    }
+
+    String supportedRegions = Arrays.stream(Regions.values()).map(Regions::getName)
+      .collect(Collectors.joining(", "));
+
+    collector.addFailure("Unsupported region value: " + region,
+      String.format("Supported regions are: %s", supportedRegions))
+      .withConfigProperty(SqsConstants.PROPERTY_REGION);
+
+    collector.getOrThrowException();
```
Comment (MEseifan) [inline src/main/java/io/cdap/plugin/aws/sqs/common/SqsPluginConfig.java:157]: `getOrThrowException()` will throw the exception if one exists. That is not recommended for methods that are called in the validation path. This `getRegion()` is called by `validateRegion()` so this should not throw an error. The collector should be used to collect all errors so they can be displayed all at once to the user.

This applies to the other get/validate methods in this class.

**Comment by shiva-cirus** [comment]
> Couple comments:
> 
> * It seems like some comments were marked as Done but nothing was changed? Maybe a git merge issue?
> * Please keep the changes in separate commits until it is time to merge. Squashing the commits during review makes it hard to see which files were updated since the last review.
> * In the last review I was wondering if there was any performance/reliability testing done, it seems like there are no additional links in the JIRA. Can we get an update on this?

Since we had to fix the google Licensing issues we decided to squash and get that out of the way. 
Limited performance testing was done and the results are in the Review doc. 

https://docs.google.com/document/d/1MMK6y6FmQiQB6Z4SV67kc3_8KQsDRceKYF3-AZK67Sg/edit#heading=h.lws3l57b46vt

**Comment by MEseifan** [comment]
Just wanted to bump this thread, there seems to still be open comments on this PR regarding unchanged files. Has that been resolved?

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java (hunk 3 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:247]: If messageFields is empty then it will never reach this line of code. This is because in case of messageFields is empty then following code will take care of that,

```
    Schema messageSchema = validateAndGetSchema(collector);
    if (messageSchema == null) {
      return;
    }
```

File: src/test/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfigTest.java (hunk 10 above)
Comment (amolkokane-cirus) [reply src/test/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfigTest.java:350]: The test case for schema with no fields already exists in the code with function name as "testValidateFormatSchemaWithNoFields"

File: src/main/java/io/cdap/plugin/aws/sqs/common/SqsPluginConfig.java (hunk 33 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/aws/sqs/common/SqsPluginConfig.java]: Done

**Comment by amolkokane-cirus** [comment]
> Just wanted to bump this thread, there seems to still be open comments on this PR regarding unchanged files. Has that been resolved?

Yes. Resolved.

File: src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java (hunk 3 above)
Comment (MEseifan) [reply src/main/java/io/cdap/plugin/aws/sqs/source/SqsSourceConfig.java:247]: I looks like the `validateAndGetSchema` function will still return a non-null value if there are no fields (see below). Is that intentional?

```
    Schema messageSchema = null;
    try {
      messageSchema = parseSchema();

      if (messageSchema.getFields().isEmpty()) {
        collector.addFailure("Schema must be specified.", null)
          .withConfigProperty(SqsSourceConstants.PROPERTY_SCHEMA);
      }
    } catch (Exception e) {...}
```

**Comment by amolkokane-cirus** [comment]
> One final comment, everything else looks good.

Done
